src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -1,21 +1,355 @@
+			sm.failSector(sectorNo)
+			newFaultCount := sm.SectorStates()[sectorNo].FaultCount
+			if newFaultCount > MAX_CONSECUTIVE_FAULT_COUNT {
+				// TODO: heavy penalization: slash pledge collateral and delete sector
+				// TODO: SendMessage(SPA.SlashPledgeCollateral)
+				sm.clearSector(sectorNo)
+			}
+		case SectorActiveStateNo:
+			// SlashStorageDealCollateral
+			// SendMessage(sma.slashStorageDealCollateral(sc.DealIDs()))
+			sm.failSector(sectorNo)
+			powerUpdate = powerUpdate - sm.Info().SectorSize()
+		case SectorFailingStateNo:
+			sm.checkMFC(sectorNo)
+		default:
+			// TODO: proper failure
+			panic("Invalid sector state in CronAction")
+		}
+	}
+
+	// Reset Proving Period and report power updates
+	// sm.ProvingPeriodEnd_ = PROVING_PERIOD_TIME
+	// SendMessage(sma.UpdatePower(powerUpdate))
+}
+
+func (sm *StorageMinerActor_I) verifyPoStSubmission(postSubmission poster.PoStSubmission) bool {
+	// 1. A proof must be submitted after the postRandomness for this proving
+	// period is on chain
+	// if rt.ChainEpoch < sm.ProvingPeriodEnd - challengeTime {
+	//   panic("too early")
+	// }
+
+	// 2. A proof must be a valid snark proof with the correct public inputs
+	// 2.1 Get randomness from the chain at the right epoch
+	// postRandomness := rt.Randomness(postSubmission.Epoch, 0)
+	// 2.2 Generate the set of challenges
+	// challenges := GenerateChallengesForPoSt(r, keys(sm.Sectors))
+	// 2.3 Verify the PoSt Proof
+	// verifyPoSt(challenges, TODO)
 
-func (sma *StorageMinerActor_I) SubmitPoSt(postProof base_mining.PoStProof, sectorStateSets sector.SectorStateSets) {
 	panic("TODO")
+	return true
 }
 
-func (sma *StorageMinerActor_I) CommitSector(onChainInfo sealing.OnChainSealVerifyInfo) {
-	currSector, found := sma.Sectors()[onChainInfo.SectorNumber()]
-	if !found {
-		sma.Sectors()[onChainInfo.SectorNumber()] = []sealing.OnChainSealVerifyInfo{onChainInfo}
+// move Sector from Active/Committed/Recovering/Failing
+// into Cleared State which means deleting the Sector from state
+// remove SectorNumber from all states on chain
+func (sm *StorageMinerActor_I) clearSector(sectorNo sector.SectorNumber) {
+	delete(sm.Sectors(), sectorNo)
+	delete(sm.SectorStates(), sectorNo)
+	sm.ProvingSet_.Remove(sectorNo)
+}
+
+// move Sector from Committed/Recovering into Active State
+// reset FaultCount to zero
+func (sm *StorageMinerActor_I) activateSector(sectorNo sector.SectorNumber) {
+	sm.SectorStates()[sectorNo] = SectorActive()
+}
+
+// move Sector from Active/Committed/Recovering into Failing State
+// and increment FaultCount
+func (sm *StorageMinerActor_I) failSector(sectorNo sector.SectorNumber) {
+	newFaultCount := sm.SectorStates()[sectorNo].FaultCount + 1
+	sm.ProvingSet_.Remove(sectorNo)
+	sm.SectorStates()[sectorNo] = SectorFailing(newFaultCount)
+}
+
+// increment FaultCount and check if greater than MaxFaultCount
+// move from Failing to Cleared State if so
+func (sm *StorageMinerActor_I) checkMFC(sectorNo sector.SectorNumber) {
```

Comment:
`MFC` in this function call should be explicit `checkMaxFaultCount`.

Also, `checkX` should not mutate state (here it does). Maybe call it `incrementFaultCounter(n sector.SectorNumber)`